File: python/leetcode/solve_n_queens.py
class Solution:
    def solveNQueens(self, n: int) -> list[list[str]]:
        results: list[list[int]] = []

        def valid(result: list[int], row: int, col: int) -> bool:
            for i, j in enumerate(result):
                if j == col or abs(i - row) == abs(j - col):
                    return False
            return True

        def next(current: list[int], row: int) -> list[list[int]]:
            if row == n:
                results.append(["." * i + "Q" + "." * (n - i - 1) for i in current])
                return
            valid_col = [col for col in range(n) if valid(current, row, col)]
            for col in valid_col:
                next(current + [col], row + 1)

        next([], 0)

        return results

File: rust/leetcode/src/solve_n_queens.rs
pub struct Solution;

impl Solution {
  pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
    fn valid(result: &Vec<i32>, row: usize, col: usize) -> bool {
      for (i, &r) in result.iter().enumerate() {
        if r == col as i32 || (r - col as i32).abs() == (i as i32 - row as i32).abs() {
          return false;
        }
      }
      true
    }

    fn next(result: &mut Vec<i32>, row: usize, n: usize, results: &mut Vec<Vec<String>>) {
      if row == n {
        results.push(
          result
            .iter()
            .map(|&c| ".".repeat(c as usize) + "Q" + &".".repeat(n - c as usize - 1))
            .collect(),
        );
        return;
      }
      for col in 0..n {
        if valid(result, row, col) {
          result.push(col as i32);
          next(result, row + 1, n, results);
          result.pop();
        }
      }
    }
    let mut results = vec![];
    let mut result = vec![];
    next(&mut result, 0 as usize, n as usize, &mut results);
    results
  }
}
